### PR TITLE
adding rosie ar kit demo

### PR DESCRIPTION
**Pull request by justin-h-m-tunis**
ported code from matlab demo into c++

### CODE REVIEW

File: util/grav_comp.hpp [hunk 1]
```diff
@@ -41,6 +41,47 @@ class GravityCompensation
...
+      {
+        wrench_vec[j] = -normed_gravity[j] * masses[i];
+      }
+
+      // Add the torques for each joint to support the mass at this frame
+      comp_torque += jacobians[i].transpose() * wrench_vec;
+    }
+        return comp_torque;
+  }
+  static Eigen::VectorXd getEfforts(
```
Comment (iamtesch) [inline util/grav_comp.hpp]: I think instead of duplicating the implementation, the implementation accepting a `GroupFeedback` can just call this implementation, right?

File: kits/rosie/arm.hpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
+#include "robot_model.hpp"
+#include "trajectory.hpp"
+
+using namespace hebi;
+
+class GripperArm{
```
Comment (iamtesch) [inline kits/rosie/arm.hpp]: I'm used to a class going into a file with the same name -- e.g., `gripper_arm.hpp` here.

I'm not sure how useful this data structure is for this one demo, but we might end up reusing it in other files.  I think the utility will go down after the arm API is finished/moved into the C++ API.

Comment (justin-h-m-tunis) [reply kits/rosie/arm.hpp]: yeah I could take it out, I don't think its super necessary either

File: util/grav_comp.hpp (hunk 1 above)
Comment (rLinks234) [reply util/grav_comp.hpp]: We need to move these out of a class. This is writing .NET/Java code here, but we are using C++. Please just make these free standing (_i.e._ outside of a class definition). I understand that you were following the existing convention, though  @justin-h-m-tunis

File: kits/rosie/arm.hpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  public:
+  std::unique_ptr<robot_model::RobotModel> model;
```
Comment (rLinks234) [inline kits/rosie/arm.hpp]: Make these private and encapsulate them

File: kits/rosie/arm.hpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+          gripper_module_name_({"Spool"}),
+	  shoulder_joint_comp_(0),
+	  effort_offset_((Eigen::VectorXd(6) << 0, shoulder_joint_comp_, 0,0,0,0).finished()),
+	  gripper_open_effort_(1),
+	  gripper_close_effort_(-5),
+	  ik_seed_pos_((Eigen::VectorXd(6) << 0,1,2.5,1.5,-1.5,1).finished()),
+	  min_traj_duration_(.50),
+	  default_speed_factor_(.9)
+	{
+    model = robot_model::RobotModel::loadHRDF("/home/hebi/hebi-cpp-examples/build/hrdf/6-dof_arm_w_gripper.hrdf");
```
Comment (rLinks234) [inline kits/rosie/arm.hpp]: This... should not be hard coded

File: kits/rosie/base.hpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
+
+#define PI 3.1415926535
+
+class Base{
+  public:
+  double wheel_radius_;
```
Comment (rLinks234) [inline kits/rosie/base.hpp]: Make private (encapsulate)

File: kits/rosie/arm.hpp (hunk 3 above)
Comment (justin-h-m-tunis) [reply kits/rosie/arm.hpp]: I also thought about using a struct to avoid like 10 different getters

Comment (rLinks234) [reply kits/rosie/arm.hpp]: Yeah... I understand your perspective here. There are a few solutions to this, but I don't think that making all fields public is the right way forward here.

Comment (justin-h-m-tunis) [reply kits/rosie/arm.hpp]: okay I'll just make them properly encapsulated the classes (if we want to keep them)

File: kits/rosie/arm.hpp (hunk 4 above)
Comment (justin-h-m-tunis) [reply kits/rosie/arm.hpp]: oh yeah I'll change this back, matt suggested I do this in order for the vscode debugger to find it

File: util/grav_comp.hpp (hunk 1 above)
Comment (iamtesch) [reply util/grav_comp.hpp]: Yeah, this is my bad here -- sorry!

File: kits/rosie/rosie_demo.cpp [hunk 6]
```diff
@@ -1,4 +1,5 @@
 #define EIGEN_INITIALIZE_MATRICES_BY_NAN
```
Comment (rLinks234) [inline kits/rosie/rosie_demo.cpp]: Why should we do this by default?

File: kits/rosie/rosie_demo.cpp [hunk 7]
```diff
@@ -77,6 +68,58 @@ void getTrajectoryCommand(std::shared_ptr<hebi::trajectory::Trajectory> traj,Gro
...
+  double a2 = 60*PI/180;
+  double a3 = 180*PI/180;
+
+  const Eigen::Matrix3d base_wheel_transform = 
+                    (Eigen::MatrixXd(3,3) <<
+                    sin(a1), -cos(a1), base_wheel_base/2,
+                    sin(a2), -cos(a2), base_wheel_base/2,
+                    sin(a3), -cos(a3), base_wheel_base/2).finished();
+
+  const Eigen::Matrix<double, 3, 3> base_wheel_velocity_matrix = base_wheel_transform / base_wheel_radius;
```
Comment (rLinks234) [inline kits/rosie/rosie_demo.cpp]: Use `Eigen::Matrix3d` here as well

File: kits/rosie/rosie_demo.cpp [hunk 8]
```diff
@@ -77,6 +68,58 @@ void getTrajectoryCommand(std::shared_ptr<hebi::trajectory::Trajectory> traj,Gro
 }
 
 int main(){
+  const std::unique_ptr<robot_model::RobotModel> arm_model = robot_model::RobotModel::loadHRDF("hrdf/6-dof_arm_w_gripper.hrdf");
+  const std::vector<std::string> arm_module_names = {"Base","Shoulder", "Elbow", "Wrist1", "Wrist2", "Wrist3"};
+  const std::vector<std::string> arm_gripper_module_name = {"Spool"};
+  const double arm_shoulder_joint_comp = 0;
+  const Eigen::Matrix<double,6,1> arm_effort_offset = (Eigen::VectorXd(6) << 0, arm_shoulder_joint_comp, 0,0,0,0).finished();
```
Comment (rLinks234) [inline kits/rosie/rosie_demo.cpp]: I'd use an alias like `Vector6d` throughout here

File: kits/rosie/rosie_demo.cpp [hunk 9]
```diff
@@ -77,6 +68,58 @@ void getTrajectoryCommand(std::shared_ptr<hebi::trajectory::Trajectory> traj,Gro
...
+  const double arm_gripper_open_effort = 1;
+  const double arm_gripper_close_effort = -5;
+  const Eigen::Matrix<double,6,1> arm_ik_seed_pos = (Eigen::VectorXd(6) << 0,1,2.5,1.5,-1.5,1).finished();
+  const double arm_min_traj_duration = .5;
+
+  const double base_wheel_radius = .15/2.0;
+  const double base_wheel_base = 0.470;
+  const double base_max_lin_speed = 0.6;
+  const double base_max_rot_speed = base_max_lin_speed*(base_wheel_base/2.0);
+  const Eigen::Matrix<double, 4, 4> base_chassis_com = (Eigen::MatrixXd(4,4) << 1,0,0,0, 0,1,0,0, 0,0,1, base_wheel_radius +.005, 0,0,0,1).finished();
```
Comment (rLinks234) [inline kits/rosie/rosie_demo.cpp]: Use `Eigen::Matrix4d` here as well

File: kits/rosie/rosie_demo.cpp [hunk 10]
```diff
@@ -239,7 +280,7 @@ int main(){
       Eigen::VectorXd masses;
       arm_model->getMasses(masses);
       Eigen::VectorXd effort(arm_dofs.size());
-      auto base_accel = fbk[base.num_wheels_ + 0].imu().accelerometer().get();
+      auto base_accel = fbk[base_num_wheels + 0].imu().accelerometer().get();
```
Comment (rLinks234) [inline kits/rosie/rosie_demo.cpp]: Why is `+ 0` here?

File: kits/rosie/rosie_demo.cpp (hunk 6 above)
Comment (justin-h-m-tunis) [reply kits/rosie/rosie_demo.cpp]: oh I dont need this anymore, I'll take it out

File: kits/rosie/rosie_demo.cpp [hunk 11]
```diff
@@ -339,9 +380,9 @@ int main(){
       }
       //
       //            % Joystick Input for Omnibase Control        
-      double x_vel = base.max_lin_speed_ * (latest_phone_fbk[0].io().a().hasInt(8) ? latest_phone_fbk[0].io().a().getInt(8) : latest_phone_fbk[0].io().a().getFloat(8)); 
-      double y_vel = -1*base.max_lin_speed_ * (latest_phone_fbk[0].io().a().hasInt(7) ? latest_phone_fbk[0].io().a().getInt(7) : latest_phone_fbk[0].io().a().getFloat(7));
-      double rot_vel = base.max_rot_speed_ * (latest_phone_fbk[0].io().a().hasInt(1) ? latest_phone_fbk[0].io().a().getInt(1) : latest_phone_fbk[0].io().a().getFloat(1)); 
+      double x_vel = base_max_lin_speed * (latest_phone_fbk[0].io().a().hasInt(8) ? latest_phone_fbk[0].io().a().getInt(8) : latest_phone_fbk[0].io().a().getFloat(8));
```
Comment (rLinks234) [inline kits/rosie/rosie_demo.cpp]: Few things here:

1) You're accessing the same fields here, so just do something like `auto& bankA = latest_phone_fbk[0].io().a();`
2) It's a bit nicer if you just have a function which corresponds to retrieving this value for you, which can also be appropriately named based on its intention. Some simple example would be:
```c++
double x_vel = base_max_lin_speed * getUserVelocityInput();
```

However... It gets messy since you have to retrieve the feedback through `latest_phone_fbk`. That indicates that some sort of structure should be created, in which the `latest_phone_fbk` is encapsulated. So this sounds out of the scope of this PR, but relevant for future PRs

File: kits/rosie/rosie_demo.cpp (hunk 10 above)
Comment (justin-h-m-tunis) [reply kits/rosie/rosie_demo.cpp]: oh it was "+1" in matlab so I just left it so that I'd be able to cross reference more easily, but i guess I can take that out now

File: kits/rosie/rosie_demo.cpp [hunk 12]
```diff
@@ -424,16 +465,16 @@ int main(){
       } else {
         t = omni_base_traj->getDuration();
       }
-      Eigen::VectorXd chassis_cmd_vel(base.num_wheels_);
-      Eigen::VectorXd chassis_cmd_acc(base.num_wheels_);
-      Eigen::VectorXd chassis_cmd_jerk(base.num_wheels_);
+      Eigen::VectorXd chassis_cmd_vel(base_num_wheels);
+      Eigen::VectorXd chassis_cmd_acc(base_num_wheels);
+      Eigen::VectorXd chassis_cmd_jerk(base_num_wheels);
```
Comment (rLinks234) [inline kits/rosie/rosie_demo.cpp]: This is very frustrating. This isn't your fault, and you're doing everything right with the current API... However, a ton of these variables in these examples have a statically set size, which will never change. Therefore, the trajectory and kinematics APIs **should** support passing in compile time sized vectors, which helps not only generate better code but also makes it harder to accidentally pass in an incorrectly sized matrix.

TL;DR: The API lacks proper functionality to allow something like `Eigen::Matrix<double, BaseNumWheels, 1> chassis_cmd_vel`, etc

File: kits/rosie/rosie_demo.cpp (hunk 11 above)
Comment (iamtesch) [reply kits/rosie/rosie_demo.cpp]: Just realizing that our about-to-be-released C++ API has a "mobile_io" class which would pretty much take care of this...

File: kits/rosie/rosie_demo.cpp [hunk 13]
```diff
@@ -0,0 +1,513 @@
...
+                     -base_accel.getZ());
+      Eigen::VectorXd arm_positions(arm_dofs.size());
+      for(size_t i = 0; i < arm_positions.size(); ++i){
+        arm_positions[i] = fbk.getPosition()[arm_dofs[i]];
+      }
+      effort = hebi::util::getGravityCompensationEfforts(*arm_model, masses, arm_positions, gravity);
+      /*TODO add dynamic_comp*/
+      setEffortSegment(cmd, arm_dofs, effort + arm_effort_offset);
+      robot_group->sendCommand(cmd);
+    }
```
Comment (iamtesch) [inline kits/rosie/rosie_demo.cpp]: Note for future refactor -- we can probably encapsulate some of these lines in this loop to reuse.

File: kits/rosie/rosie_demo.cpp [hunk 14]
```diff
@@ -0,0 +1,513 @@
...
+    Eigen::MatrixXd accelerations = Eigen::MatrixXd::Zero(3,2); 
+    Eigen::MatrixXd jerks = Eigen::MatrixXd::Zero(3,2); 
+    
+    auto omni_base_traj = hebi::trajectory::Trajectory::createUnconstrainedQp(time, velocities, &accelerations, &jerks);
+    bool first_run = true;
+    
+    while(true){
+      //            %%%%%%%%%%%%%%%%%%%
+      //            % Gather Feedback %
+      //            %%%%%%%%%%%%%%%%%%%
```
Comment (iamtesch) [inline kits/rosie/rosie_demo.cpp]: Note for future tidy/cleanup -- these are definitely MATLAB comments that were commented out in C++ :)   We can make this look more standard by just having these be a C++ comment characters, possibly with left alignment

File: kits/rosie/rosie_demo.cpp [hunk 15]
```diff
@@ -168,35 +136,35 @@ int main(){
      
   auto wheel_group = lookup.getGroupFromNames(robot_family, base_wheel_module_names);
   GroupCommand base_gains_command(wheel_group->size());
-  if (!base_gains_command.readGains("gains/omni-drive-wheel-gains.xml")){
+  if (!base_gains_command.readGains("gains/omni-drive-wheel-gains.xml")) {
     printf("Could not read omni base gains");
-    return -1;
+    return 1;
```
Comment (rLinks234) [inline kits/rosie/rosie_demo.cpp]: Process return codes should never be negative

File: util/grav_comp.hpp [hunk 16]
```diff
@@ -1,66 +1,56 @@
...
-    {
-      // Set translational part
-      for (size_t j = 0; j < 3; ++j)
-      {
-        wrench_vec[j] = -normed_gravity[j] * masses[i];
-      }
-
-      // Add the torques for each joint to support the mass at this frame
-      comp_torque += jacobians[i].transpose() * wrench_vec;
+inline Eigen::VectorXd getGravityCompensationEfforts(const hebi::robot_model::RobotModel& model,
```
Comment (rLinks234) [inline util/grav_comp.hpp]: `inline` modifier still makes the function "publicly" available (default visibility as opposed to static linkage) in each translation unit, but marking it as `inline` specifies to the linker that only one implementation needs to be chosen at link time. So technically, if you have two different implementations of function `foo`, which are both marked inline, it's not defined which version will be linked in. but you're doing weird stuff to begin with, if you do that...
